src/p11_cert.c

```diff
@@ -295,6 +299,33 @@ int pkcs11_store_certificate(PKCS11_TOKEN *token, X509 *x509, char *label,
 	pkcs11_addattr_int(attrs + n++, CKA_CERTIFICATE_TYPE, CKC_X_509);
 	pkcs11_addattr_obj(attrs + n++, CKA_SUBJECT,
 		(pkcs11_i2d_fn)i2d_X509_NAME, X509_get_subject_name(x509));
+    pkcs11_addattr_obj(attrs + n++, CKA_ISSUER,
+        (pkcs11_i2d_fn)i2d_X509_NAME, X509_get_issuer_name(x509));
+
+    /* Get digest algorithm from x509 certificate */
+    evp_md = EVP_get_digestbynid(X509_get_signature_nid(x509));
+    switch (EVP_MD_type(evp_md)) {
+    case NID_sha1:
+        ckm_md = CKM_SHA_1;
+    case NID_sha224:
+        ckm_md = CKM_SHA224;
+    case NID_sha256:
+        ckm_md = CKM_SHA256;
+    case NID_sha512:
+        ckm_md = CKM_SHA512;
+    case NID_sha384:
+        ckm_md = CKM_SHA384;
+    default:
+        ckm_md = 0;
+    }
+
+    /* Set hash algorithm; default is SHA-1 */
```

Comment:
if SHA-1 is the default, then please make that explicit in the switch statement above, i.e. move `default:` above `case NID_sha1:`.

Reply:
Actually SHA-1 is the default CKA_NAME_HASH_ALGORITHM.
Meaning that if ckm_md is equal to 0, SHA-1 will be use. This comes from PKCS11.